// cypress-fail-fast/dist/Browser/FailFast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerFailFast = registerFailFast;
const Constants_1 = require("../Shared/Constants");
const Config_1 = require("../Shared/Config");
const CypressHelpers_1 = require("./CypressHelpers");
/**
 * Registers Mocha hooks that implement fail-fast behavior in the browser process.
 * @param Cyp Cypress global object.
 * @param cy Cypress command API.
 * @param before Global `before` hook.
 * @param beforeEach Global `beforeEach` hook.
 * @param afterEach Global `afterEach` hook.
 */
function registerFailFast(Cyp, cy, before, beforeEach, afterEach) {
    /**
     * Logs a message through the Node-side logging task.
     * @param message Message to log.
     */
    function log(message) {
        cy.task(Constants_1.LOG_TASK, message);
    }
    /**
     * Reads the global skip flag from Node tasks.
     * @returns Cypress chainable resolving to skip state.
     */
    function shouldSkip() {
        return cy.task(Constants_1.SHOULD_SKIP_TASK, undefined, { log: false });
    }
    /**
     * Resets the global skip flag.
     */
    function resetSkipFlag() {
        cy.task(Constants_1.RESET_SKIP_TASK, null, { log: false });
    }
    /**
     * Resets the global failed-tests counter.
     */
    function resetFailedTests() {
        cy.task(Constants_1.RESET_FAILED_TESTS_TASK, null, { log: false });
    }
    /**
     * Enables skip mode for subsequent tests.
     * @returns Cypress chainable resolving when skip mode is enabled.
     */
    function enableSkipMode(failedTest) {
        log(Constants_1.SKIP_MESSAGE);
        return cy.task(Constants_1.TRIGGER_FAIL_FAST_TASK, {
            test: failedTest,
        });
    }
    function mapFailedTest(currentTest) {
        return {
            name: currentTest.title,
            fullTitle: currentTest.fullTitle(),
        };
    }
    /**
     * Increments the failed-tests counter.
     * @returns Cypress chainable resolving to total failed tests.
     */
    function registerFailure() {
        return cy.task(Constants_1.FAILED_TESTS_TASK, true, { log: false });
    }
    /**
     * Runs a callback once the configured bail threshold is reached.
     * @param callback Callback executed when failed tests reach bail limit.
     */
    function registerFailureAndRunIfBailLimitIsReached(callback) {
        registerFailure().then((value) => {
            const bail = (0, Config_1.bailConfig)(Cyp);
            log(`${Constants_1.FAILED_TEST_MESSAGE}: ${value}/${bail}`);
            if (value >= bail) {
                callback();
            }
        });
    }
    /**
     * Runs a callback only when skip mode is active.
     * @param callback Callback to execute in skip mode.
     */
    function runIfSkipIsEnabled(callback) {
        shouldSkip().then((value) => {
            if (value === true) {
                callback();
            }
        });
    }
    /**
     * Skips the current suite when plugin and skip mode are both enabled.
     * @param context Mocha execution context.
     */
    function skipSuiteIfEnabled(context) {
        runIfSkipIsEnabled(() => {
            context.skip();
        });
    }
    before(function () {
        if ((0, CypressHelpers_1.isHeaded)(Cyp) || (0, Config_1.currentStrategyIsSpec)(Cyp)) {
            /*
              Reset the shouldSkip flag at the start of a run, so that it doesn't carry over into subsequent runs. Do this only for headed runs because in headless runs, the `before` hook is executed for each spec file.
            */
            resetSkipFlag();
            resetFailedTests();
        }
        else {
            skipSuiteIfEnabled(this);
        }
    });
    beforeEach(function () {
        skipSuiteIfEnabled(this);
    });
    afterEach(function () {
        // Mark skip flag as true if test failed
        const currentTest = this.currentTest;
        if (currentTest &&
            (0, CypressHelpers_1.testHasFailed)(currentTest) &&
            (0, CypressHelpers_1.failFastIsEnabled)(currentTest, Cyp)) {
            log(`Test "${currentTest.fullTitle()}" has failed`);
            registerFailureAndRunIfBailLimitIsReached(() => {
                enableSkipMode(mapFailedTest(currentTest));
            });
        }
    });
}
