// cypress-fail-fast/dist/Shared/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RUN_STRATEGY = exports.SPEC_STRATEGY = void 0;
exports.isTruthy = isTruthy;
exports.isFalsy = isFalsy;
exports.isUndefined = isUndefined;
exports.strategyIsSpec = strategyIsSpec;
exports.strategyValue = strategyValue;
exports.getFailFastEnvironmentConfig = getFailFastEnvironmentConfig;
exports.getFailFastPluginConfig = getFailFastPluginConfig;
exports.currentStrategyIsSpec = currentStrategyIsSpec;
exports.shouldIgnorePerTestConfig = shouldIgnorePerTestConfig;
exports.bailConfig = bailConfig;
const Constants_1 = require("./Constants");
exports.SPEC_STRATEGY = "spec";
exports.RUN_STRATEGY = "run";
const truthyValuesSet = new Set([true, "true", 1, "1"]);
const falsyValuesSet = new Set([false, "false", 0, "0"]);
const undefinedValuesSet = new Set(["", undefined]);
/**
 * Determines whether a value should be interpreted as truthy in env parsing.
 * @param value Value to evaluate.
 * @returns `true` for accepted truthy values.
 */
function isTruthy(value) {
    return truthyValuesSet.has(value);
}
/**
 * Determines whether a value should be interpreted as falsy in env parsing.
 * @param value Value to evaluate.
 * @returns `true` for accepted falsy values.
 */
function isFalsy(value) {
    return falsyValuesSet.has(value);
}
/**
 * Determines whether a value is considered undefined for env parsing.
 * @param value Value to evaluate.
 * @returns `true` for empty string or `undefined`.
 */
function isUndefined(value) {
    return undefinedValuesSet.has(value);
}
/**
 * Checks if the provided strategy value equals `spec`.
 * @param value Strategy value from Cypress env.
 * @returns `true` when strategy is `spec`.
 */
function strategyIsSpec(value) {
    return value === exports.SPEC_STRATEGY;
}
/**
 * Normalizes a strategy value to one of the supported strategy constants.
 * @param value Strategy value to normalize.
 * @returns Normalized strategy value.
 */
function strategyValue(value) {
    return strategyIsSpec(value) ? exports.SPEC_STRATEGY : exports.RUN_STRATEGY;
}
function isDefined(value) {
    return !isUndefined(value);
}
function booleanVarValue(value, defaultValue) {
    const isTruthyValue = isTruthy(value);
    if (!isTruthyValue && !isFalsy(value)) {
        return defaultValue;
    }
    return isTruthyValue;
}
function numericVarValue(value, defaultValue) {
    if (isDefined(value)) {
        return Number(value);
    }
    return defaultValue;
}
/**
 * Builds the effective fail-fast configuration from Cypress env variables.
 * @param Cyp Cypress global object.
 * @returns Parsed fail-fast global configuration.
 */
function getFailFastEnvironmentConfig(Cyp) {
    const strategy = strategyValue(Cyp.expose(Constants_1.STRATEGY_GLOBAL_CONFIG));
    return {
        strategy,
        ignorePerTestConfig: booleanVarValue(Cyp.expose(Constants_1.IGNORE_PER_TEST_CONFIG), Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.IGNORE_PER_TEST_CONFIG]),
        enabled: booleanVarValue(Cyp.expose(Constants_1.ENABLED_GLOBAL_CONFIG), Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.ENABLED_GLOBAL_CONFIG]),
        strategyIsSpec: strategy === exports.SPEC_STRATEGY,
        bail: numericVarValue(Cyp.expose(Constants_1.BAIL_GLOBAL_CONFIG), Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.BAIL_GLOBAL_CONFIG]),
    };
}
/**
 * Builds the effective fail-fast configuration from resolved plugin config values.
 * @param config Cypress plugin configuration.
 * @returns Parsed fail-fast global configuration.
 */
function getFailFastPluginConfig(config) {
    const strategy = strategyValue(config.expose?.[Constants_1.STRATEGY_GLOBAL_CONFIG]);
    return {
        strategy,
        ignorePerTestConfig: booleanVarValue(config.expose?.[Constants_1.IGNORE_PER_TEST_CONFIG], Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.IGNORE_PER_TEST_CONFIG]),
        enabled: booleanVarValue(config.expose?.[Constants_1.ENABLED_GLOBAL_CONFIG], Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.ENABLED_GLOBAL_CONFIG]),
        strategyIsSpec: strategy === exports.SPEC_STRATEGY,
        bail: numericVarValue(config.expose?.[Constants_1.BAIL_GLOBAL_CONFIG], Constants_1.GLOBAL_CONFIG_DEFAULT_VALUES[Constants_1.BAIL_GLOBAL_CONFIG]),
    };
}
/**
 * Returns whether the current fail-fast strategy is `spec`.
 * @param Cyp Cypress global object.
 * @returns `true` when strategy is `spec`.
 */
function currentStrategyIsSpec(Cyp) {
    return getFailFastEnvironmentConfig(Cyp).strategyIsSpec;
}
/**
 * Returns whether the configuration per-test overrides should be ignored in favor of global configuration.
 * @param Cyp Cypress global object.
 * @returns `true` when per-test configuration should be ignored.
 */
function shouldIgnorePerTestConfig(Cyp) {
    return getFailFastEnvironmentConfig(Cyp).ignorePerTestConfig;
}
/**
 * Returns the configured bail threshold.
 * @param Cyp Cypress global object.
 * @returns Number of failures required to start skipping tests.
 */
function bailConfig(Cyp) {
    return getFailFastEnvironmentConfig(Cyp).bail;
}
