// cypress-fail-fast/dist/Shared/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL_CONFIG_DEFAULT_VALUES = exports.LOG_PREFIX = exports.FAILED_TEST_MESSAGE = exports.SKIP_MESSAGE = exports.RESET_FAILED_TESTS_TASK = exports.FAILED_TESTS_TASK = exports.LOG_TASK = exports.RESET_SKIP_TASK = exports.TRIGGER_FAIL_FAST_TASK = exports.SHOULD_SKIP_TASK = exports.BAIL_GLOBAL_CONFIG = exports.STRATEGY_GLOBAL_CONFIG = exports.ENABLED_GLOBAL_CONFIG = exports.IGNORE_PER_TEST_CONFIG = void 0;
/** Global config key to ignore per-test fail-fast configuration and use only global config. */
exports.IGNORE_PER_TEST_CONFIG = "failFastIgnorePerTestConfig";
/** Global config that enables or disables fail-fast behavior. */
exports.ENABLED_GLOBAL_CONFIG = "failFastEnabled";
/** Global config that selects fail-fast strategy. */
exports.STRATEGY_GLOBAL_CONFIG = "failFastStrategy";
/** Global config that defines the failure threshold. */
exports.BAIL_GLOBAL_CONFIG = "failFastBail";
/** Task name to read current skip state. */
exports.SHOULD_SKIP_TASK = "failFastShouldSkip";
/** Task name to enable skip mode with fail-fast metadata. */
exports.TRIGGER_FAIL_FAST_TASK = "failFastTrigger";
/** Task name to reset skip state. */
exports.RESET_SKIP_TASK = "failFastResetSkip";
/** Task name to write plugin logs in Node process. */
exports.LOG_TASK = "failFastLog";
/** Task name to increment/read failed tests counter. */
exports.FAILED_TESTS_TASK = "failFastFailedTests";
/** Task name to reset failed tests counter. */
exports.RESET_FAILED_TESTS_TASK = "failFastResetFailedTests";
/** Message emitted when skip mode becomes active. */
exports.SKIP_MESSAGE = "Enabling skip mode";
/** Prefix used for failed tests progress logs. */
exports.FAILED_TEST_MESSAGE = "Failed tests";
/** Prefix used in console log lines emitted by this plugin. */
exports.LOG_PREFIX = "[fail-fast]";
/** Default values used when related fail-fast global config variables are not provided. */
exports.GLOBAL_CONFIG_DEFAULT_VALUES = {
    [exports.IGNORE_PER_TEST_CONFIG]: false,
    [exports.ENABLED_GLOBAL_CONFIG]: true,
    [exports.BAIL_GLOBAL_CONFIG]: 1,
    [exports.STRATEGY_GLOBAL_CONFIG]: "run",
};
